File: Python/hopcroft_karp.py
from collections import deque

# BFS to construct the level graph
def bfs_level_graph(graph, pair_u, pair_v, dist):
   queue = deque()
   for u in range(len(graph)):
      if pair_u[u] == -1:
         dist[u] = 0
         queue.append(u)
      else:
         dist[u] = float('inf')
   dist[-1] = float('inf')

   while queue:
      u = queue.popleft()
      if dist[u] < dist[-1]:
         for v in graph[u]:
            if dist[pair_v[v]] == float('inf'):
               dist[pair_v[v]] = dist[u] + 1
               queue.append(pair_v[v])

   return dist[-1] != float('inf')

# DFS to find augmenting paths
def dfs_augmenting_path(graph, pair_u, pair_v, dist, u):
   if u != -1:
      for v in graph[u]:
         if dist[pair_v[v]] == dist[u] + 1:
            if dfs_augmenting_path(graph, pair_u, pair_v, dist, pair_v[v]):
               pair_v[v] = u
               pair_u[u] = v
               return True
      dist[u] = float('inf')
      return False
   return True

# Main function to implement the Hopcroft-Karp algorithm
def hopcroft_karp(graph):
   pair_u = [-1] * len(graph)
   pair_v = [-1] * len(graph[0])
   dist = [-1] * len(graph)
   matching = 0

   while bfs_level_graph(graph, pair_u, pair_v, dist):
      for u in range(len(graph)):
         if pair_u[u] == -1:
            if dfs_augmenting_path(graph, pair_u, pair_v, dist, u):
               matching += 1
   return matching

# Example graph for testing the Hopcroft-Karp algorithm
graph = [
   [0, 1, 1, 0],
   [1, 0, 0, 1],
   [0, 1, 0, 1],
   [1, 0, 1, 0]
]

# Output the maximum matching
print("Maximum Matching:", hopcroft_karp(graph))

File: TypeScript/hopcroft_karp.ts
/**
 * BFS to construct the level graph
 */
function bfsLevelGraph(
  graph: number[][],
  pairU: number[],
  pairV: number[],
  dist: Map<number, number>
): boolean {
  const queue: number[] = [];

  for (let u = 0; u < graph.length; u++) {
    if (pairU[u] === -1) {
      dist.set(u, 0);
      queue.push(u);
    } else {
      dist.set(u, Infinity);
    }
  }

  dist.set(-1, Infinity);

  while (queue.length > 0) {
    const u = queue.shift()!;

    if (dist.get(u)! < dist.get(-1)!) {
      for (const v of graph[u]) {
        if (dist.get(pairV[v]) === Infinity) {
          dist.set(pairV[v], dist.get(u)! + 1);
          queue.push(pairV[v]);
        }
      }
    }
  }

  return dist.get(-1) !== Infinity;
}

/**
 * DFS to find augmenting paths
 */
function dfsAugmentingPath(
  graph: number[][],
  pairU: number[],
  pairV: number[],
  dist: Map<number, number>,
  u: number
): boolean {
  if (u !== -1) {
    for (const v of graph[u]) {
      if (dist.get(pairV[v]) === dist.get(u)! + 1) {
        if (dfsAugmentingPath(graph, pairU, pairV, dist, pairV[v])) {
          pairV[v] = u;
          pairU[u] = v;
          return true;
        }
      }
    }
    dist.set(u, Infinity);
    return false;
  }
  return true;
}

/**
 * Main function to implement the Hopcroft-Karp algorithm
 */
function hopcroftKarp(graph: number[][]): number {
  const pairU = new Array(graph.length).fill(-1);
  const pairV = new Array(graph[0].length).fill(-1);
  const dist = new Map<number, number>();
  let matching = 0;

  while (bfsLevelGraph(graph, pairU, pairV, dist)) {
    for (let u = 0; u < graph.length; u++) {
      if (pairU[u] === -1) {
        if (dfsAugmentingPath(graph, pairU, pairV, dist, u)) {
          matching++;
        }
      }
    }
  }
  return matching;
}

// Example usage:
const graph4 = [
  [1, 2],    // Node 0 connects to 1, 2
  [0, 3],    // Node 1 connects to 0, 3
  [1, 3],    // Node 2 connects to 1, 3
  [0, 2]     // Node 3 connects to 0, 2
];

console.log("Maximum Matching:", hopcroftKarp(graph4));
